presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
- have you noticed that `Lookup::resolve` got deprecated? We should try to avoid calling this. This is a last resort, to be used only if there is no other way.
 - I still do believe it should be a separate rule. Notice that we sort `WindowNode`s in the same way (`SwapAdjacentWindowsBySpecifications`), just to make them eligible to be merged. `WindowNode` sorting on its own has not much sense. This is why I suggested to put all the rules related to `WindowNode`s merging into the same `RuleSet`. To emphasis that these rules work in cooperation.
 - IMO it does not conflict with unused column pruning rules as they both push down the projection.